sdk/attestation/azure-security-attestation/azure/security/attestation/_client.py

```diff
@@ -133,50 +172,112 @@ def attest_sgx_enclave(self, quote, inittime_data=None, runtime_data=None, **kwa
                 :dedent: 8
                 :caption: Attesting an SGX Enclave
 
-        For additional request configuration options, please see `Python Request Options <https://aka.ms/azsdk/python/options>`_.
+        For additional request configuration options, please see `Python Request
+        Options <https://aka.ms/azsdk/python/options>`_.
 
         """
+
+        inittime_json = kwargs.pop("inittime_json", None)  # type: bytes
+        inittime_data = kwargs.pop("inittime_data", None)  # type: bytes
+        runtime_json = kwargs.pop("runtime_json", None)  # type: bytes
+        runtime_data = kwargs.pop("runtime_data", None)  # type: bytes
+
+        if inittime_json and inittime_data:
+            raise ValueError("Cannot provide both inittime_json and inittime_data.")
+        if runtime_json and runtime_data:
+            raise ValueError("Cannot provide both runtime_data and runtime_json.")
+
+        # If the input was JSON, make sure that it's valid JSON before sending it
```

Comment:
Is this client-side validation necessary? I assume so since it's here, but what happens if the JSON is invalid when sent?

Reply:
The server will fail the request if you specify JSON and it's not valid JSON. But that check happens fairly late in the attestation process (after validating the attestation evidence).

But as a matter of general principle, my philosophy is that it's better to check for errors on the client rather than the server if it's an easy check. And parsing JSON is an easy check IMHO.

It's a bit of a LBYL attitude, but when dealing with server interactions, I believe it's better to check on the client than churn network bandwidth (and potentially costing the customer $$$ - the attestation service is free, but other Azure services are not).

Reply:
According to the [Python guidelines](https://azure.github.io/azure-sdk/python_design.html#parameter-validation), service parameters shouldn't be validated in any case (we should just work with the service to make errors helpful). It seems like there's at least some layer of client-side logic here though, so I think validation is permissible if you wouldn't consider this a direct service parameter

Reply:
It 100% isn't a direct service parameter. Based on the inittime_data/inittime_json parameters, we then encapsulate those in an AttestationData model type and then encapsulate _that_ in an AttestationRequest model type. _That_ is the direct service parameter.

My philosophy here has always been to ensure that the data being sent to the service is as valid as possible because checks on the client are orders of magnitude cheaper than checks on the service. So if the client hands us a PEM encoded X.509 certificate, we verify that the PEM value actually contains an X.509 certificate and is valid. And then we generated the JWK needed by the service from the decoded certificate. This does two things: 1) it verifies that the certificate is valid and 2) Ensures that the value sent to the service is canonical (PEM decode is mechanical, but there are a lot of variables and parsing logic that I don't want to have to deal with - I'd rather let some other library do that work).